modules/distributor/queue/queue.go

```diff
@@ -0,0 +1,224 @@
+	}, []string{"name", "tenant"})
+	pushesFailuresTotalMetric := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_pushes_failures_total",
+		Help:      "Total number of failed pushes to the queue for a tenant to the generatorForwarder",
+	}, []string{"name", "tenant"})
+	lengthMetric := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_length",
+		Help:      "Number of queued requests for a tenant",
+	}, []string{"name", "tenant"})
+
+	var alreadyRegisteredErr prometheus.AlreadyRegisteredError
+	err := reg.Register(pushesTotalMetrics)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
+		_ = level.Warn(logger).Log("msg", "failed to register queue_pushes_total metric", "err", err)
+	}
+	err = reg.Register(pushesFailuresTotalMetric)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
+		_ = level.Warn(logger).Log("msg", "failed to register queue_pushes_failures_total metric", "err", err)
+	}
+
+	err = reg.Register(lengthMetric)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
+		_ = level.Warn(logger).Log("msg", "failed to register queue_length metric", "err", err)
+	}
+
+	return &Queue[T]{
+		logger:                     logger,
+		name:                       cfg.Name,
+		tenantID:                   cfg.TenantID,
+		workerCount:                cfg.WorkerCount,
+		size:                       cfg.Size,
+		reqChan:                    make(chan *request[T], cfg.Size),
+		fn:                         fn,
+		workersCloseCh:             make(chan struct{}),
+		pushesTotalMetrics:         pushesTotalMetrics,
+		pushesFailuresTotalMetrics: pushesFailuresTotalMetric,
+		lengthMetric:               lengthMetric,
+		readOnly:                   atomic.NewBool(false),
+	}
+}
+
+// Push pushes data onto a queue.
+// If the queue is full, the data is dropped
+func (m *Queue[T]) Push(ctx context.Context, data T) error {
+	if m.readOnly.Load() {
+		return fmt.Errorf("queue is read-only")
+	}
+
+	m.pushesTotalMetrics.WithLabelValues(m.name, m.tenantID).Inc()
+
+	req := &request[T]{
+		data: data,
+	}
+
+	select {
+	case <-ctx.Done():
+		m.pushesFailuresTotalMetrics.WithLabelValues(m.name, m.tenantID).Inc()
+		return fmt.Errorf("failed to push data to queue for tenant=%s and queue_name=%s: %w", m.tenantID, m.name, ctx.Err())
+	default:
+	}
+
+	select {
+	case m.reqChan <- req:
+		m.lengthMetric.WithLabelValues(m.name, m.tenantID).Inc()
+		return nil
+	default:
+	}
+
+	// Fail fast if the queue is full
+	m.pushesFailuresTotalMetrics.WithLabelValues(m.name, m.tenantID).Inc()
+
+	return fmt.Errorf("failed to push data to queue for tenant=%s and queue_name=%s: queue is full", m.tenantID, m.name)
+}
+
+func (m *Queue[T]) StartWorkers() {
+	for i := 0; i < m.workerCount; i++ {
+		m.wg.Add(1)
+
+		go m.worker()
+	}
+}
+
+// Name returns queue name.
+func (m *Queue[T]) Name() string {
```

Comment:
nit: if these are used then keep them, otherwise delete them.

no reason to have accessors that are unused.